src/OpenSage.Mods.Generals/Logic/OrderGenerators/GeneralsOrderGenerator.cs

```diff
@@ -0,0 +1,492 @@
+
+    public GeneralsOrderGenerator(Game game)
+    {
+        _game = game;
+    }
+
+    public void BuildRenderList(RenderList renderList, Camera camera, in TimeInterval gameTime) { }
+
+    private string GetCursorForForceFire()
+    {
+        if (SelectedUnitsIsStructure(out var s) && s.CanAttack)
+        {
+            return GetCursorForForceAttackingStructure(s);
+        }
+
+        if (_worldObject is null)
+        {
+            return GeneralsCursors.ForceAttackGround;
+        }
+
+        if (SelectedUnits.Count == 1 && SelectedUnits.Single() == _worldObject)
+        {
+            return GeneralsCursors.Select;
+        }
+
+        return AnyUnitCanAttackTarget(_worldObject) ? GeneralsCursors.AttackObj : GeneralsCursors.GenericInvalid;
+    }
+
+    private static string GetCursorForWaypointModifier()
+    {
+        // pretty much just always return waypoint as long as we own the unit selected
+        return GeneralsCursors.Waypoint;
+    }
+
+    public string GetCursor(KeyModifiers keyModifiers)
+    {
+        var selectedUnits = _game.Scene3D.LocalPlayer?.SelectedUnits;
+        if (selectedUnits is not null && (selectedUnits.Count == 0 ||
+                                          selectedUnits.All(u => u.Owner != LocalPlayer)))
+        {
+            return _worldObject != null
+                ? GeneralsCursors.Select // TODO: Maybe shouldn't have this here.
+                : GeneralsCursors.Arrow;
+        }
+
+        // the local player has selected unit(s), and they are the owner of those unit(s)
+
+        if (keyModifiers.HasFlag(KeyModifiers.Alt))
+        {
+            return GetCursorForWaypointModifier();
+        }
+
+        if (keyModifiers.HasFlag(KeyModifiers.Ctrl))
+        {
+            return GetCursorForForceFire();
+        }
+
+        // no modifier key has been applied
+        // SelectedUnits is player-owned
+
+        // TODO: verify a garrisoned civilian structure is in fact player-owned
+        if (SelectedUnitsIsStructure(out var s))
+        {
+            return GetCursorForStructureSelected(s);
+        }
+
+        // the selected object is a player-owned unit (not structure)
+        if (!TerrainUnderTargetIsRevealed())
+        {
+            // In Zero Hour, if terrain is under fog of war we always return the Move cursor, no matter what
+            // arguably we could improve upon this by looking at the terrain under the cursor, and ignoring any world objects
+            return GeneralsCursors.Move;
+        }
+
+        // area is not under fog of war
+        if (!TryGetTarget(out var target))
+        {
+            return GetCursorForTerrainTarget();
+        }
+
+        // target is some game object
+        if (TargetIsEnemy(target))
+        {
+            return GetCursorForEnemyTarget(target);
+        }
+
+        // target is not enemy
+        if (AnySelectedUnitCanCaptureTarget(target))
+        {
+            return GeneralsCursors.CaptureBuilding;
```

Comment:
I was just digging into the ini files some more and noticed that for `CommandButton Command_AmericaRangerCaptureBuilding` there is a specified `CursorName` as well as `InvalidCursorName`. Special powers, fire weapon commands, etc should probably use whatever is defined there instead of being hardcoded here.